packages/next/client/components/app-router.client.tsx

```diff
@@ -91,6 +91,25 @@ let initialParallelRoutes: CacheNode['parallelRoutes'] =
 
 const prefetched = new Set<string>()
 
+/**
+ * The pathname and search params are stored in the leaf node of the router state tree.
+ */
+function getPathnameAndSearchParams(tree: FlightRouterState): {
+  pathname: string
+  searchParams: { [key: string]: string | string[] }
+} {
+  const pathname = tree[4]
+  const searchParams = tree[5]
+
+  if (pathname && searchParams) {
+    return { pathname, searchParams }
+  }
+
+  const parallelRoutes = tree[1]
+  // TODO-APP: pick the correct parallel route based on marker
+  return getPathnameAndSearchParams(Object.values(parallelRoutes)[0])
```

Comment:
Reading the first item here is not enough, it needs to iterate all branches to see if any of them hold the pathname/searchParams. Given that these are related maybe it should be a single item `tree[4]` that holds `[pathname, searchParams]`? 🤔 
```suggestion
  return getPathnameAndSearchParams(Object.values(parallelRoutes)[0])
```